shell/platform/darwin/ios/framework/Source/FlutterPlatformViews.mm

```diff
@@ -655,96 +668,227 @@ static bool ClipRRectContainsPlatformViewBoundingRect(const SkRRect& clip_rrect,
+    // If frame is null, AcquireFrame already printed out an error message.
+    if (!frame) {
+      continue;
+    }
+    DlCanvas* overlay_canvas = frame->Canvas();
+    int restore_count = overlay_canvas->GetSaveCount();
+    overlay_canvas->Save();
+    overlay_canvas->ClipRect(overlay->second);
+    overlay_canvas->Clear(DlColor::kTransparent());
+    slices_[view_id]->render_into(overlay_canvas);
+    overlay_canvas->RestoreToCount(restore_count);
+
+    // This flutter view is never the last in a frame, since we always submit the
+    // underlay view last.
+    frame->set_submit_info(
+        {.frame_boundary = false,
+         .present_with_transaction = true,
+         .submit_receiver = [&callbacks](const SurfaceFrame::DeferredSubmit& cb) {
+           callbacks.push_back(cb);
+         }});
+
+    layer->did_submit_last_frame = frame->Submit();
+
     did_submit &= layer->did_submit_last_frame;
-    platform_view_layers[view_id].push_back(layer);
+    platform_view_layers[view_id].push_back(
+        {.rect = overlay->second, .view_id = view_id, .overlay_id = overlay_id, .layer = layer});
     overlay_id++;
   }
 
-  // If a layer was allocated in the previous frame, but it's not used in the current frame,
-  // then it can be removed from the scene.
-  RemoveUnusedLayers();
-  // Organize the layers by their z indexes.
-  BringLayersIntoView(platform_view_layers);
+  background_frame->set_submit_info(
+      {.present_with_transaction = true,
+       .submit_receiver = [&callbacks](const SurfaceFrame::DeferredSubmit& cb) {
+         callbacks.push_back(cb);
+       }});
+  did_submit &= background_frame->Submit();
+
   // Mark all layers as available, so they can be used in the next frame.
+  std::vector<std::shared_ptr<FlutterPlatformViewLayer>> unused_layers =
+      layer_pool_->RemoveUnusedLayers();
   layer_pool_->RecycleLayers();
 
-  did_submit &= frame->Submit();
+  auto task = [&, platform_view_layers = std::move(platform_view_layers),  //
+               current_composition_params = current_composition_params_,   //
+               views_to_recomposite = views_to_recomposite_,               //
+               callbacks = std::move(callbacks),                           //
+               composition_order = composition_order_,                     //
+               unused_layers = std::move(unused_layers),
+               views_to_dispose = DisposeViews()  //
+  ]() mutable {
+    PerformSubmit(platform_view_layers,        //
+                  callbacks,                   //
+                  current_composition_params,  //
+                  views_to_recomposite,        //
+                  composition_order,           //
+                  unused_layers,               //
+                  views_to_dispose             //
+    );
+  };
+
+  // Workaround for FlutterPlatformViewsTest.mm
+  if ([[NSThread currentThread] isMainThread]) {
```

Comment:
`RunNowOrPostTask` instead? Its hard to fathom what the workaround is for.

Reply:
I should update the comment, its for iOS simualtors with software backend. The trick to split across threads doesn't work there so I still need to merge threads.

Yeah, will switch to RunNowOrPostTask.

Reply:
And elsewhere.